### PR TITLE
Stabilize dask loading of Ocean Model data

### PR DESCRIPTION
**Pull request by CKehl**
This PR is a clone or mirror of #774 , which suffers from a broken CI.

The lazy evaluation from Dask is still lacking robustness and some users still experience errors in some setups, which this PR attempts to resolve.

Test are developed for NEMO, GlobCurrent and POP.

### CODE REVIEW

File: parcels/examples/example_dask_chunk_OCMs.py [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+    dimensions = {'U': {'lon': 'glamf', 'lat': 'gphif', 'depth': 'depthw', 'time': 'time_counter'},
+                  'V': {'lon': 'glamf', 'lat': 'gphif', 'depth': 'depthw', 'time': 'time_counter'},
+                  'W': {'lon': 'glamf', 'lat': 'gphif', 'depth': 'depthw', 'time': 'time_counter'}}
+    chs = False
+    if chunk_mode == 'auto':
+        chs = 'auto'
+    elif chunk_mode == 'specific':
+        chs = {'U': {'depthu': 75, 'y': 16, 'x': 16},
+               'V': {'depthv': 75, 'y': 16, 'x': 16},
+               'W': {'depthw': 75, 'y': 16, 'x': 16}}
```
Comment (erikvansebille) [inline parcels/examples/example_dask_chunk_OCMs.py]: Would it not be a better test to make these chunking dictionaries different? That would test the code even more?

File: parcels/examples/example_dask_chunk_OCMs.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+
+def compute_pop_particle_advection(field_set, mode, lonp, latp):
+    pset = ParticleSet.from_list(field_set, ptype[mode], lon=lonp, lat=latp)
+    pfile = ParticleFile("globcurrent_particles_chunk", pset, outputdt=delta(days=15))
+    pset.execute(AdvectionRK4, runtime=delta(days=90), dt=delta(days=2), output_file=pfile)
+    return pset
+
+
+@pytest.mark.parametrize('mode', ['jit'])  # Only testing jit as scipy is very slow
+@pytest.mark.parametrize('chunk_mode', [False, 'auto', 'specific'])  # Only testing jit as scipy is very slow
```
Comment (erikvansebille) [inline parcels/examples/example_dask_chunk_OCMs.py]: Don't need the comment here (it applies to line above)?

File: parcels/examples/example_dask_chunk_OCMs.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+    assert (len(field_set.U.grid.load_chunk) == len(field_set.W.grid.load_chunk))
+    if chunk_mode is False:
+        assert (len(field_set.U.grid.load_chunk) == 1)
+    elif chunk_mode == 'auto':
+        assert (len(field_set.U.grid.load_chunk) != 1)
+    elif chunk_mode == 'specific':
+        assert (len(field_set.U.grid.load_chunk) == (1 * int(math.ceil(201.0/16.0)) * int(math.ceil(151.0/16.0))))
+
+
+# ==== undefined as long as we have no POP example data ==== #
```
Comment (erikvansebille) [inline parcels/examples/example_dask_chunk_OCMs.py]: Remove this comment?

File: parcels/examples/example_dask_chunk_OCMs.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+@pytest.mark.parametrize('mode', ['jit'])  # Only testing jit as scipy is very slow
+@pytest.mark.parametrize('chunk_mode', [False, 'auto', 'specific'])  # Only testing jit as scipy is very slow
```
Comment (erikvansebille) [inline parcels/examples/example_dask_chunk_OCMs.py]: Again, remove the comment here?

File: parcels/examples/example_dask_chunk_OCMs.py [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+    if chunk_mode is False:
+        assert (len(field_set.U.grid.load_chunk) == 1)
+    elif chunk_mode == 'auto':
+        assert (len(field_set.U.grid.load_chunk) != 1)
+    elif chunk_mode == 'specific':
+        assert (len(field_set.U.grid.load_chunk) == (int(math.ceil(21.0/3.0)) * int(math.ceil(60.0/8.0)) * int(math.ceil(60.0/8.0))))
+
+
+@pytest.mark.parametrize('mode', ['jit'])  # Only testing jit as scipy is very slow
+@pytest.mark.parametrize('chunk_mode', [False, 'auto', 'specific'])  # Only testing jit as scipy is very slow
```
Comment (erikvansebille) [inline parcels/examples/example_dask_chunk_OCMs.py]: And here again stray comment

File: parcels/examples/example_dask_chunk_OCMs.py [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+    elif chunk_mode == 'auto':
+        assert (len(field_set.U.grid.load_chunk) != 1)
+    elif chunk_mode == 'specific':
+        assert (len(field_set.U.grid.load_chunk) == (1 * int(math.ceil(41.0/16.0)) * int(math.ceil(81.0/16.0))))
+    assert(abs(pset[0].lon - 23.8) < 1)
+    assert(abs(pset[0].lat - -35.3) < 1)
+
+
+@pytest.mark.parametrize('mode', ['jit'])  # Only testing jit as scipy is very slow
+def test_diff_entry_dimensions_chunks(mode):
```
Comment (erikvansebille) [inline parcels/examples/example_dask_chunk_OCMs.py]: How is this function different from the `test_nemo_3D()` above?

File: parcels/examples/example_nemo_curvilinear.py [hunk 7]
```diff
@@ -101,6 +104,102 @@ def test_nemo_3D_samegrid():
...
+    pset = ParticleSet.from_list(field_set, ptype[mode], lon=lonp, lat=latp)
+    pfile = ParticleFile("nemo_particles", pset, outputdt=delta(days=1))
+    kernels = pset.Kernel(AdvectionRK4) + periodicBC
+    pset.execute(kernels, runtime=delta(days=4), dt=delta(hours=6),
+                 output_file=pfile, recovery={ErrorCode.ErrorOutOfBounds: OutOfBounds_reinitialisation})
+    return pset
+
+
+@pytest.mark.parametrize('mode', ['jit'])  # Only testing jit as scipy is very slow
+def test_nemo_curvilinear_auto_chunking(mode):
```
Comment (erikvansebille) [inline parcels/examples/example_nemo_curvilinear.py]: Why are these functions below here and not in the `example_dask_chunk_OCMs.py` file?

File: parcels/examples/example_dask_chunk_OCMs.py (hunk 6 above)
Comment (CKehl) [reply parcels/examples/example_dask_chunk_OCMs.py]: Look at the dimensions - here, we test that we do not chunk the depth dimension, even if it's in the chunk dictionary, as long as it is not also listed in the `dimensions` dictionary. Those two do have some affect on one another.

File: parcels/examples/example_nemo_curvilinear.py (hunk 7 above)
Comment (CKehl) [reply parcels/examples/example_nemo_curvilinear.py]: hm - so you prefer to take them out here ? In the end, I think the `test_nemo_3D` include those tests anyway.

Comment (CKehl) [reply parcels/examples/example_nemo_curvilinear.py]: removed the functions as they were a carbon-copy of example_dask_chunk_OCMs.py for NEMO indeed. Thanks for that feedback!

File: parcels/examples/example_dask_chunk_OCMs.py [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+    chs = {'U': {'depthu': 75, 'y': 16, 'x': 16},
+           'V': {'depthv': 20, 'y': 4, 'x': 16},
+           'W': {'depthw': 15, 'y': 16, 'x': 4}}
+    fieldset = FieldSet.from_nemo(filenames, variables, dimensions, field_chunksize=chs)
+    npart = 20
+    lonp = 5.2 * np.ones(npart)
+    latp = [i for i in 52.0+(-1e-3+np.random.rand(npart)*2.0*1e-3)]
+    try:
+        compute_nemo_particle_advection(fieldset, mode, lonp, latp)
+    except:
```
[the remark this answers is not in the document]
Comment (CKehl) [reply parcels/examples/example_dask_chunk_OCMs.py]: yes, good point - I'll adapt that.

File: parcels/field.py [hunk 9]
```diff
@@ -901,8 +904,8 @@ def get_block(self, bid):
         return np.unravel_index(bid, self.nchunks[1:])
 
     def chunk_setup(self):
-        if self.chunk_set:
-            return
+        #if self.chunk_set:
+        #    return
```
[the remark this answers is not in the document]
Comment (CKehl) [reply parcels/field.py]: yes

File: parcels/examples/example_dask_chunk_OCMs.py [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+    npart_V = [npart_V * k for k in fieldset.V.nchunks[1:]]
+    npart_V_request = 1
+    chn = {'U': {'lat': int(math.ceil(41.0/chs['U']['lat'])),
+                 'lon': int(math.ceil(81.0/chs['U']['lon']))},
+           'V': {'lat': int(math.ceil(41.0/chs['V']['lat'])),
+                 'lon': int(math.ceil(81.0/chs['V']['lon']))}}
+    npart_V_request = [npart_V_request * chn['V'][k] for k in chn['V']]
+    assert (npart_U == npart_V)
+    assert (npart_V != npart_V_request)
+    assert (len(fieldset.U.grid.load_chunk) == len(fieldset.V.grid.load_chunk))
```
[the remark this answers is not in the document]
Comment (CKehl) [reply parcels/examples/example_dask_chunk_OCMs.py]: oops - yes

File: parcels/fieldset.py [hunk 11]
```diff
@@ -274,39 +276,56 @@ def from_netcdf(cls, filenames, variables, dimensions, indices=None, fieldtype=N
...
+                            else:
+                                res |= (chunksize == grid.master_chunksize)
+                            if res:
+                                grid_chunksize = grid.master_chunksize
+                                logger.warning("Trying to initialize a shared grid with different chunking sizes - action prohibited. Replacing requested field_chunksize with grid's master chunksize.")
+                            else:
+                                raise ValueError("Conflict between grids of the same fieldset chunksize and requested field chunksize as well as the chunked name dimensions - Please apply the same chunksize to all fields in a shared grid!")
+                        if procpaths == nowpaths:
+                            dFiles = fields[procvar].dataFiles
+                            break
```
Comment (erikvansebille) [inline parcels/fieldset.py:315]: Is this the best place to do this check? What if a user first creates a `FieldSet.from_netcdf()` and then later adds a `Field` via `FieldSet.add_field()`? This check then wouldn't be done. 
Would it make sense to move/combine this check with the new one in gridset.py, which is always called when a new `Field` is added to a `FieldSet`?

Comment (CKehl) [reply parcels/fieldset.py:315]: I think that is so and so, can be argued. At this place, we know we treat NetCDF file data, so their grids need to have ``field_chunksize`` parameters. We also know which other fields and grids are there.

In the ``add_field()`` function we also catch later-added Fields, but there, we have the messy task to check if that Field is a netcdf field, if it has a field_chunksize, etc.

I'll have a look how to integrate this as nicely as possible.

File: parcels/examples/example_dask_chunk_OCMs.py [hunk 12]
```diff
@@ -0,0 +1,435 @@
...
+    compute_nemo_particle_advection(fieldset, mode, lonp, latp)
+    # Nemo sample file dimensions: depthu=75, y=201, x=151
+    assert (len(fieldset.U.grid.load_chunk) == len(fieldset.V.grid.load_chunk))
+
+
+# ==== TO BE EXTERNALIZED OR CHECKED WHEN #782 IS FIXED ==== #
+@pytest.mark.parametrize('mode', ['scipy', 'jit'])
+def test_3d_2dfield_sampling(mode):
+    logger.warning("Test is to be re-enabled after #782 is fixed to test tertiary effects.")
+    return True
```
[the remark this answers is not in the document]
Comment (CKehl) [reply parcels/examples/example_dask_chunk_OCMs.py:213]: thanks for the hint!

File: parcels/examples/example_dask_chunk_OCMs.py [hunk 13]
```diff
@@ -0,0 +1,435 @@
...
+    fieldset.add_field(Field('rectilinear_2D', np.ones((2, 2)),
+                             lon=np.array([-10, 20]), lat=np.array([40, 80]), field_chunksize=False))
+
+    class MyParticle(ptype[mode]):
+        sample_var_curvilinear = Variable('sample_var_curvilinear')
+        sample_var_rectilinear = Variable('sample_var_rectilinear')
+    pset = ParticleSet(fieldset, pclass=MyParticle, lon=2.5, lat=52)
+
+    def Sample2D(particle, fieldset, time):
+        print("Interpolating curvilinear grid - sample=%.04f." % (particle.sample_var_curvilinear))
```
[the remark this answers is not in the document]
Comment (CKehl) [reply parcels/examples/example_dask_chunk_OCMs.py:240]: to be cleaned when this function actually works verifiably. (i.e. after 782)

File: parcels/examples/example_dask_chunk_OCMs.py [hunk 14]
```diff
@@ -0,0 +1,435 @@
...
+        particle.sample_var_curvilinear += fieldset.nav_lon[time, particle.depth, particle.lat, particle.lon]
```
[the remark this answers is not in the document]
Comment (CKehl) [reply parcels/examples/example_dask_chunk_OCMs.py:241]: to be re-organised when this function actually works verifiably. (i.e. after 782)

File: parcels/examples/example_dask_chunk_OCMs.py [hunk 15]
```diff
@@ -0,0 +1,435 @@
...
+    for f in fieldset.get_fields():
+        if type(f) in [VectorField, NestedField, SummedField]:  # or not f.grid.defer_load:
+            continue
+        g = f.grid
+        npart = 1
+        npart = [npart * k for k in f.nchunks[1:]]
+        print("Field '{}': grid type: {}; grid chunksize: {}; grid mesh: {}; field N partitions: {}; field nchunks: {}; grid chunk_info: {}; grid load_chunk: {}; grid layout: {}".format(f.name, g.gtype, g.master_chunksize, g.mesh, npart, f.nchunks, g.chunk_info, g.load_chunk, (g.tdim, g.zdim, g.ydim, g.xdim)))
+    for i in range(0, len(fieldset.gridset.grids)):
+        g = fieldset.gridset.grids[i]
+        print("Grid {}: grid type: {}; grid chunksize: {}; grid mesh: {}; grid chunk_info: {}; grid load_chunk: {}; grid layout: {}".format(i, g.gtype, g.master_chunksize, g.mesh, g.chunk_info, g.load_chunk, (g.tdim, g.zdim, g.ydim, g.xdim)))
```
[the remark this answers is not in the document]
Comment (CKehl) [reply parcels/examples/example_dask_chunk_OCMs.py:259]: to be cleaned when this function actually works verifiably. (i.e. after #782 )

File: parcels/fieldset.py (hunk 11 above)
Comment (CKehl) [reply parcels/fieldset.py:315]: working on it

Comment (CKehl) [reply parcels/fieldset.py]: the change is now done, as discussed

**Comment by CKehl** [comment]
The code runs all tests fine, but the test_palaeo_dataset.py still fails.

**Pull request merged**